### PR TITLE
fix: stabilize nats connection

### PR DESCRIPTION
**Pull request by yhl25**
We will be completely relying on NATs auto-reconnection as ours have potential data race bugs.

* introduce client pool
  * reuse the same clients based on a static pool size
* removed the jetstream client interface and also our auto-reconnection
* refactored the nats client to reorganize the code
* removed proxy interface
* manually implement heartbeat for watchers

TODO:
unit tests - I will create a separate pr for that

### CODE REVIEW

File: pkg/sources/source.go [hunk 1]
```diff
@@ -60,12 +60,16 @@ func (sp *SourceProcessor) Start(ctx context.Context) error {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 	var writersMap = make(map[string][]isb.BufferWriter)
-
+	// FIXME: make size configurable
+	natsClientPool, err := jsclient.NewClientPool(ctx, jsclient.WithClientPoolSize(2))
```
Comment (vigith) [inline pkg/sources/source.go:64]: isn't the default pool size of 3 fine?

File: pkg/udf/map_udf.go [hunk 2]
```diff
@@ -44,11 +45,19 @@ type MapUDFProcessor struct {
...
 	log := logging.FromContext(ctx)
+	finalWg := sync.WaitGroup{}
+
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	fromBuffer := u.VertexInstance.Vertex.OwnedBuffers()
-	finalWg := sync.WaitGroup{}
 
+	//FIXME: make size configurable
+	natsClientPool, err := jsclient.NewClientPool(ctx, jsclient.WithClientPoolSize(3))
```
Comment (vigith) [inline pkg/udf/map_udf.go:54]: leave it at default?

File: pkg/udf/reduce_udf.go [hunk 3]
```diff
@@ -52,10 +53,24 @@ type ReduceUDFProcessor struct {
...
+		err            error
+		natsClientPool *jsclient.ClientPool
+		windower       window.Windower
+	)
 	log := logging.FromContext(ctx)
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	var windower window.Windower
+
+	//FIXME: make size configurable
```
Comment (vigith) [inline pkg/udf/reduce_udf.go:68]: 3?

File: pkg/watermark/publish/publisher.go [hunk 4]
```diff
@@ -176,7 +176,7 @@ func (p *publish) validateWatermark(wm wmb.Watermark, toVertexPartitionIdx int32
 		p.log.Debugw("New watermark is updated for the head watermark", zap.String("head", headWM.String()), zap.String("new", wm.String()))
 		p.SetHeadWM(wm, toVertexPartitionIdx)
 	} else if wm.Before(time.Time(headWM)) {
-		p.log.Warnw("Skip publishing the new watermark because it's older than the current watermark", zap.String("entity", p.entity.GetName()), zap.Int64("head", headWM.UnixMilli()), zap.Int64("new", wm.UnixMilli()))
+		p.log.Debugw("Skip publishing the new watermark because it's older than the current watermark", zap.String("entity", p.entity.GetName()), zap.Int64("head", headWM.UnixMilli()), zap.Int64("new", wm.UnixMilli()))
```
Comment (vigith) [inline pkg/watermark/publish/publisher.go:179]: please do not move this to Debug unless we have a metric to track it.

**Review by vigith — APPROVED** [review]
with NITs

File: pkg/watermark/store/jetstream/kv_watch.go [hunk 5]
```diff
@@ -123,68 +113,113 @@ func (jsw *jetStreamWatch) Watch(ctx context.Context) (<-chan store.WatermarkKVE
...
 					}
-				case nats.KeyValuePurge:
-					// do nothing
 				}
+			case <-jsw.kvwTimer.C:
+				// if the timer expired, it means that the watcher is not receiving any updates
+				kvLastUpdatedTime := jsw.lastUpdateKVTime()
+				// if the last update time is before the previous fetch time, it means that the store is not getting any updates
+				// therefore, we don't have to recreate the watcher
+				if kvLastUpdatedTime.Before(jsw.previousFetchTime) {
```
Comment (whynowy) [inline pkg/watermark/store/jetstream/kv_watch.go:164]: good idea.

**Review by whynowy — APPROVED** [review]

**Pull request merged**